src/core/or/circuitpadding.c

```diff
@@ -0,0 +1,2370 @@
+     * this node. */
+    circpad_cell_event_nonpadding_sent(circ);
+  }
+}
+
+/**
+ * Free the machineinfo at an index
+ */
+static void
+circpad_circuit_machineinfo_free_idx(circuit_t *circ, int idx)
+{
+  if (circ->padding_info[idx]) {
+    tor_free(circ->padding_info[idx]->histogram);
+    timer_free(circ->padding_info[idx]->padding_timer);
+    tor_free(circ->padding_info[idx]);
+  }
+}
+
+/**
+ * Free all padding machines and mutable info associated with
+ * circuit
+ */
+void
+circpad_circuit_machineinfo_free(circuit_t *circ)
+{
+  FOR_EACH_CIRCUIT_MACHINE_BEGIN(i) {
+    circpad_circuit_machineinfo_free_idx(circ, i);
+  } FOR_EACH_CIRCUIT_MACHINE_END;
+}
+
+/**
+ * Allocate a new mutable machineinfo structure.
+ */
+circpad_machineinfo_t *
+circpad_circuit_machineinfo_new(circuit_t *on_circ, int machine_index)
+{
+  circpad_machineinfo_t *mi = tor_malloc_zero(sizeof(circpad_machineinfo_t));
+  mi->machine_index = machine_index;
+  mi->on_circ = on_circ;
+
+  return mi;
+}
+
+static void
+circpad_setup_machine_on_circ(circuit_t *on_circ, circpad_machine_t *machine)
+{
+  if (CIRCUIT_IS_ORIGIN(on_circ) && !machine->origin_side) {
+    log_fn(LOG_WARN, LD_BUG,
+           "Can't set up non-origin machine on origin circuit!");
+    return;
+  }
+
+  if (!CIRCUIT_IS_ORIGIN(on_circ) && machine->origin_side) {
+    log_fn(LOG_WARN, LD_BUG,
+           "Can't set up origin machine on non-origin circuit!");
+    return;
+  }
+
+  tor_assert_nonfatal(on_circ->padding_machine[machine->machine_index]
+                      == NULL);
+  tor_assert_nonfatal(on_circ->padding_info[machine->machine_index] == NULL);
+
+  on_circ->padding_info[machine->machine_index] =
+      circpad_circuit_machineinfo_new(on_circ, machine->machine_index);
+  on_circ->padding_machine[machine->machine_index] = machine;
+}
+
+static void
+circpad_circ_client_machine_init(void)
+{
+  circpad_machine_t *circ_client_machine
+      = tor_malloc_zero(sizeof(circpad_machine_t));
+
+  // XXX: Better conditions for merge.. Or disable this machine in
+  // merge?
+  circ_client_machine->conditions.min_hops = 2;
+  circ_client_machine->conditions.state_mask =
+      CIRCPAD_CIRC_BUILDING|CIRCPAD_CIRC_OPENED|CIRCPAD_CIRC_HAS_RELAY_EARLY;
+  circ_client_machine->conditions.purpose_mask = CIRCPAD_PURPOSE_ALL;
+
+  circ_client_machine->target_hopnum = 2;
+  circ_client_machine->origin_side = 1;
+
+  circ_client_machine->start.transition_events[CIRCPAD_STATE_BURST] =
+    CIRCPAD_EVENT_NONPADDING_RECV;
+
+  circ_client_machine->burst.transition_events[CIRCPAD_STATE_BURST] =
+    CIRCPAD_EVENT_PADDING_RECV |
+    CIRCPAD_EVENT_NONPADDING_RECV;
+
+  /* If we are in burst state, and we send a non-padding cell, then we cancel
+     the timer for the next padding cell:
+     We dont want to send fake extends when actual extends are going on */
+  circ_client_machine->burst.transition_cancel_events =
+    CIRCPAD_EVENT_NONPADDING_SENT;
+
+  circ_client_machine->burst.transition_events[CIR
```

Comment:
Can we split the parts of this file that do padding negotiation into their own module? This file is pretty large...

Reply:
I opened a ticket about this in https://trac.torproject.org/projects/tor/ticket/28777#ticket because I think trying to split the file now will be pretty hasty and won't give us a good interface. It will also cause lots of conflicts since me and Mike are both working on the same branch. Let's do it after the branch is merged IMO.